File: LeetCode/1110. Delete Nodes And Return Forest/1110.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def delNodes(self, root, to_delete):
        """
        :type root: TreeNode
        :type to_delete: List[int]
        :rtype: List[TreeNode]
        """
        to_delete = set(to_delete)
        res = []

        def dfs(node, is_root):
            if not node:
                return None
            if node.val in to_delete:
                node.left = dfs(node.left, True)
                node.right = dfs(node.right, True)
                return None
            else:
                if is_root:
                    res.append(node)
                node.left = dfs(node.left, False)
                node.right = dfs(node.right, False)
                return node

        dfs(root, True)
        return res
        
        #other people's solution
        '''
        class Solution(object):
    def delNodes(self, root, to_delete):
        """
        :type root: TreeNode
        :type to_delete: List[int]
        :rtype: List[TreeNode]
        """
        if not root:
            return []
        result = []
        to_delete_set = set(to_delete)
        self.dfs(root, to_delete_set, result)
        if root.val not in to_delete_set:
            result.append(root)
        return result
        
    def dfs(self, node, to_delete_set, result):
        if not node:
            return None
        node.left = self.dfs(node.left, to_delete_set, result)
        node.right = self.dfs(node.right, to_delete_set, result)
        if node.val in to_delete_set:
            if node.left:
                result.append(node.left)
            if node.right:
                result.append(node.right)
            return None
        else:
            return node
        '''

File: LeetCode/1110. Delete Nodes And Return Forest/1110.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        List<TreeNode> res = new ArrayList<>();
        Set<Integer> set = new HashSet<>();
        for (int i : to_delete) {
            set.add(i);
        }
        dfs(root, set, res, true);
        return res;
        
    }

    public TreeNode dfs(TreeNode root, Set<Integer> set, List<TreeNode> res, boolean isRoot) {
        if (root == null) {
            return null;
        }
        boolean deleted = set.contains(root.val);
        if (isRoot && !deleted) {
            res.add(root);
        }
        root.left = dfs(root.left, set, res, deleted);
        root.right = dfs(root.right, set, res, deleted);
        return deleted ? null : root;
    }

}